### PR TITLE
logtail/filch: limit buffer file size to 50MB

### PR DESCRIPTION
**Pull request by maisem**
If for some reason there is no network connectivity, the log files keep growing.  This attempts to limit the max size we consume to 50MB per file.

image

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: logtail/filch/filch.go [hunk 1]
```diff
@@ -91,6 +93,17 @@ func (f *Filch) scan() ([]byte, error) {
 func (f *Filch) Write(b []byte) (int, error) {
 	f.mu.Lock()
 	defer f.mu.Unlock()
+	fi, err := f.cur.Stat()
```
Comment (crawshaw) [inline logtail/filch/filch.go:96]: it would be nice to avoid the stat on every write. put a log counter in Filch, n and check every 100 writes?

**Review by crawshaw — APPROVED** [review]
I think this works. Any chance to write a test with maxFileSize dialed way down?

File: logtail/filch/filch.go (hunk 1 above)
Comment (josharian) [reply logtail/filch/filch.go:96]: Or go stateless and do checks randomly. Or put a size counter in Filch, at which point you don't need a stat ever.

**Comment by maisem** [comment]
> I think this works. Any chance to write a test with maxFileSize dialed way down?

Added a test to verify that we can only read maxFileSize x 2 bytes.

**Review by crawshaw — APPROVED** [review]
LGTM, thanks!

**Pull request merged**